src/ray/design_docs/id_specification.md

```diff
@@ -0,0 +1,59 @@
+                                                                          +-----------------+
+                                                                          |   unique bytes  |  JobID     4B
+                                                                          +-----------------+
+
+                                                                 4B                4B
+                                                        +-----------------+-----------------+
+                                                        |   random bytes  |       JobID     |  ActorID   8B
+                                                        +-----------------+-----------------+
+
+                                        6B                               8B
+                            +---------------------------+-----------------------------------+
+                            |       random bytes        |               ActorID             |  TaskID   14B
+                            +---------------------------+-----------------------------------+
+
+       4B            2B                                  14B
++---------------------------+---------------------------------------------------------------+
+|  index bytes  |flags bytes|                             TaskID                            |  ObjectID 20B
++---------------------------+---------------------------------------------------------------+
+
+```
+#### JobID
+`JobID` is generated by `GCS` to ensure uniqueness. Its length is 4 bytes.
+
+#### ActorID
+An `ActorID` contains 2 parts: one is a 4 random bytes and one is the bytes of its `JobID`.
+
+#### TaskID
+A `TaskID` contains 2 parts as well: one is a 6 random bytes and one is the bytes of `ActorID` to which it belongs.
+
+#### ObjectID
+An `ObjectID` contains 3 parts:
+- `index bytes`: 4 bytes to indicate the index of the object.
+- `flags bytes`: 2 bytes to indicate the flags of this object. We have 3 flags now: `is_task`, `object_type` and `transport_type`.
+- `TaskID`: 14 bytes to indicate the id of the task to which this object belongs.
+
+**flags bytes format**
+```
+  1b     1b        3b                          11b
++-------------------------------------------------------------------------+
+| (1) | (2) |     (3)      |                (4)unused                     |
++-------------------------------------------------------------------------+
+```
+- The (1) `is_task` part is one bit to indicat whether this object id generated from a task.
+
+- The (2) `object_type` part is one bit to indicate the type of this object, whether a `PUT_OBJECT` or a `RETURN_OBJECT`.
+    - `PUT_OBJECT` indicates this object is generated through `ray.put`.
+    - `RETURN_OBJECT` indicates this object is the return value of a task.
+
+- The (3) `transport_type` part is 3 bits to indicate the type of the transport which is used to transfer this object. So it can support 8 typess.
```

Comment:
As discussed with @raulchen and @zhijunfu on this thread https://github.com/ray-project/ray/pull/5286#discussion_r307952893 , we'd like to ` GetTransportType` returns a `uint8_t`, because `id` module shouldn't know the details of transport type.